### ISSUE TITLE
SizedRange and is_sized_range

### ISSUE THREAD

**Issue by CaseyCarter** [ericniebler/stl2#67]
If I want to use my range type `R` which meets the syntactic requirements of but does not model `SizedRange` with the standard library, am I required to specialize `is_sized_range<R>` to `false_type`? Or am I required to specialize `is_sized_range<U>` to `false_type` for all `U` such that `is_same_v<uncvref_t<U>, R>` is `true`? The current wording would seem to indicate the latter.

I think `is_sized_range` should be insensitive to cv-quals/references for user convenience. That makes implementation of the default complicated since `Range` and hence `SizedRange` are sensitive to cv-quals/references. It's more easily implemented as:
```
namespace detail {
template <class T>
concept bool SizedRangeLike =
  Range<T>() &&
  requires (T& t) {
    { size(t) } -> Integral; // non-modifying
    requires ConvertibleTo<decltype(size(t)), DifferenceType<IteratorType<T>>>();
  };  
}

template <class>
struct is_sized_range : std::true_type {};

template <class T>
concept bool SizedRange() {
  return meta::_v<is_sized_range<uncvref_t<T>>> &&
    detail::SizedRangeLike<T>;
}
```
Thoughts?

**Comment by ericniebler** [comment]
I prefer an implementation like [that in range-v3](https://github.com/ericniebler/range-v3/blob/master/include/range/v3/range_concepts.hpp#L418). `is_sized_range<X>` and `is_sized_range<X const &>` should return the same result always, and they should always default to `false` if `X` is not a range.

```c++
template <class X>
struct __is_sized_range : std::false_type {};

template <class T, std::size_t N>
struct __is_sized_range<T[N]> : std::true_type {};

template <__SizedRangeLike X>
struct __is_sized_range<X> : std::true_type {};

template <class T>
concept bool _Unqualified = Same<T, __uncvref_t<T>>();

template <class T>
struct is_sized_range : is_sized_range<__uncvref_t<T>> {};

template <_Unqualified T>
struct is_sized_range<T> : __is_sized_range<T> {};
```

**Comment by ericniebler** [comment]
Implemented in  6411f9a

**Comment by CaseyCarter** [comment]
This indirectly adds a constraint that `SizedRange<T>` requires `SizedRangeLike<uncvref_t<T>>` unless `is_sized_range<uncvref_t<T>>` is specialized, making the special case for array types necessary. This is almost certainly the case for views, but I can imagine a paranoid container implementer doing something special with rvalue containers like maybe returning `dangling` from `begin` and `end`. That additional requirement doesn't provide any functionality: `is_sized_range` is not (and shouldn't be) useful in any context outside of `SizedRange`. Why add a constraint and expend one more template instantiation and two more concept checks if it doesn't buy anything?

**Comment by ericniebler** [comment]
> `is_sized_range` is not (and shouldn't be) useful in any context outside of `SizedRange`.

I disagree. Traits are usable in contexts where concepts are not. If we leave a trait like `is_sized_range` lying about, you can be certain it will show up in code. For it to return `true` for types that are not even `Range`s is going to catch people out. We can give it sensible behavior, so we should.

**Comment by CaseyCarter** [comment]
This prevents a range `R` that is immutable-only - provides only `const` flavors of `begin` and `end` - from modeling `SizedRange<const R&>` unless it specializes `is_sized_range<R>` to `true_type`. This seems strange to me.

**Comment by CaseyCarter** [comment]
> We can give it sensible behavior, so we should.

No, we can't. `is_sized_range<R>::value == false` tells me that `R` does not model `SizedRange`, `is_sized_range<R>::value == true` tells me that either `R` meets the syntactic requirements of `SizedRangeLike` or that someone specialized it to `true`. It notably does not indicate that `R` models `SizedRange`. If we want to provide a trait that is `true` if and only if `R` models `SizedRange` we could do so - but `is_sized_range` is not that trait.

**Comment by ericniebler** [comment]
This seems to be an argument that `is_sized_range` is misnamed. Perhaps we can think of a name that more clearly communicates that it is *only* for determining conformance to the `SizedRange` concept.

One possible solution is the much-despised traits blob. Like `iterator_traits`, we can have a `range_traits` and put `is_view` and `is_sized_range` there as class statics. Thoughts?

**Comment by CaseyCarter** [comment]
Meh. I too hate the traits blob. Mixing viewy stuff into rangey stuff feels wrong to me. Also, I'd be just as concerned about mis-use of `range_traits<R>::is_XXX` as I would about `is_XXX<R>`. I think abandoning the connotation-heavy name `is_XXX` would be best. `opt_out_of_XXX` is horrible but at least clearly indicates "THIS IS NOT A USEFUL TRAIT" right on the label. `disable_XXX` might be a bit better?

**Comment by ericniebler** [comment]
> Mixing viewy stuff into rangey stuff feels wrong to me.

I don't understand this. There are two refinements of `Range`, both of which could be satisfied "accidentally": `View` and `SizedRange`. The two traits are orthogonal.

> I'd be just as concerned about mis-use of `range_traits<R>::is_XXX` as I would about `is_XXX<R>`.

It's actually better than that. Consider:

```c++
template <Range R>
struct range_traits {
  constexpr static bool is_view = /* logic */;
  constexpr static bool is_sized = /* logic */;
};
```

Users can't specialize `range_traits` on types that aren't ranges.

In fact, we could even use this to make the `is_sized_range` trait safer. What if `is_sized_range` were defined like this:

```c++
template <__SizedRangeLike R>
struct is_sized_range : std::true_type {};
```

That seems to solve the problem, doesn't it?

**Comment by ericniebler** [comment]
This seems to work:

```c++
namespace detail {
template <class T>
concept bool _SizedRangeLike =
  Range<T&>() &&
  requires (T& t) {
    STL2_DEDUCTION_CONSTRAINT(size(t), Integral);
    STL2_CONVERSION_CONSTRAINT(size(t), DifferenceType<IteratorType<T&>>);
  };
template <class T>
concept bool _Unqual = Same<T, uncvref_t<T>>();
}

template <detail::_SizedRangeLike T>
struct is_sized_range : is_sized_range<detail::uncvref_t<T>> {};

template <detail::_SizedRangeLike T>
  requires detail::_Unqual<T>
struct is_sized_range<T> : std::true_type {};

template <class T>
concept bool SizedRange() {
  return detail::_SizedRangeLike<T>
    && meta::_v<is_sized_range<T>>;
}
```

**Comment by CaseyCarter** [comment]
> That seems to solve the problem, doesn't it?

That's better. It has a precondition that `Range<R>` is satisfied if `R` meets the syntactic requirements of `Range`, which could be clearer with more syntax:
```
template <Range R>
struct is_sized_range : false_type {};
template <__SizedRangeLike R>
struct is_sized_range<R> : true_type {};
```
in conjunction with the as-yet-unwritten wording about UB resulting from instantiating templates with parameters that meet the syntax but not the semantics of concept constraints.

It notably still has the issue that immutable-only ranges cannot satisfy SizedRange without specializing `is_sized_range`.

I'm also inclined to think we should specify that the check of the opt-out trait happens before the check of _SizedRangeLike in SizedRange, in case `size(t)` is a hard error?

**Comment by CaseyCarter** [comment]
> Or perhaps it's already implied...

Yes, it's already implied. So that formulation of `is_sized_range` is well-defined.

**Comment by CaseyCarter** [comment]
> hard error? (e.g., `int size() const = delete`)
Doesn't cause a hard error, but this does:
```
template <typename T = void*>
decltype(std::declval<T>() - std::declval<T>()) size() const;
```

**Comment by ericniebler** [comment]
> It notably still has the issue that immutable-only ranges cannot satisfy `SizedRange` without specializing `is_sized_range`.

Does it? Why?

```c++
template <typename T = void*>
decltype(std::declval<T>() - std::declval<T>()) size() const;
```

It does, but that seems surprising to me, and the error I get makes no sense (invalid use of `void`). It's possible this is simply a compiler bug. Regardless, it hardly seems like an important case.

**Comment by CaseyCarter** [comment]
> Does it? Why?

The check for Range<T&>() in SizedRangeLike<T> will fail.

All of this complexity is a result of violating the single responsibility principle since `is_sized_range` is both a mechanism to opt out of `SizedRange` and a means to test if some type satisfies `SizedRange`. I think it greatly simplifies the design to separate those two responsibilities into a distinct opt-out trait and a satisfaction predicate. To wit:
```
template <class>
struct disable_sized_range : std::false_type {};

template <class T>
concept bool SizedRange() {
  return !disable_sized_range<detail::uncvref_t<T>>::value &&
    detail::SizedRangeLike<T>;
}

template <class>
struct is_sized_range : std::false_type {};
template <SizedRange R>
struct is_sized_range<R> : std::true_type {};
```
This provides a useful `is_sized_range` - Bikeshed: I think I'd prefer a variable template `IsSizedRange` or `models::SizedRange` - allows for a single point of customization, and works with immutable-only containers and mutable-only views.

That "single point of customization" concern has me thinking that your `range_traits<R>` suggestion has merit. It would be easier to specialize `range_traits<R>` than to separately specialize `disable_sized_range` and/or `disable_view`.

**Comment by ericniebler** [comment]
I have it implemented over here. This range:

```c++
struct sized_range {
  int const* begin() const;
  int const* end() const;
  int size() const;
};
```

passes these tests:

```c++
  CHECK(models::sized_range<sized_range>());
  CHECK(models::sized_range<sized_range &>());
  CHECK(models::sized_range<sized_range const>());
  CHECK(models::sized_range<sized_range const &>());
```

Why do you think an immutable-only range fails the `Range` test? Shall I commit this so you can try it yourself?

**Comment by CaseyCarter** [comment]
> Why do you think an immutable-only range fails the Range test?

Because I somehow convinced myself that `begin(std::declval<T&>())` would fail if `T` has only a `const` overload of `begin`. We'll just pretend that never happened. The only remaining issue I see is if instantiating the declaration of `size` causes a hard error - which is admittedly a very small one - and I think it will go away if we decide to go the `range_traits` route. Go ahead and commit what you have to cmcstl2; I'll make a wording change for STL2 and we'll get this issue closed.

**Comment by ericniebler** [comment]
> We'll just pretend that never happened.

LOL

> I think it will go away if we decide to go the `range_traits` route.

I'm less enamored of this solution than I am with the separate `is_sized_range` trait. With the trait blob, it's possible for `range_traits<R>::is_sized` to be true even if `size(r)` doesn't compile. With the separate trait, that's not the case. And I'd like to investigate whether the hard error is actually a compiler bug. I can't make sense of it.

**Comment by CaseyCarter** [comment]
GCC hard errors `decltype(std::declval<T*>() - std::declval<T*>())` where
`T` is an incomplete type, including `void`. I stumbled across it when
writing `DifferenceType` - hence the specialization of
`DifferenceType<void*>`. At first I thought it was a bug, but I think that
hard errors are appropriate instead of SFINAE when the substitution depends
on the completeness of a type. Not sure what the standard says.

That particular hard error isn't likely to arise in practice. It was the
first thing that came to mind that would cause a hard error instantiating
the declaration of a template member function. I'd like to enable users to
adapt third-party ranges for use with the standard library even if those
ranges have an insane member size, but it's clearly not an issue of great
importance.

On Tue, Jul 28, 2015 at 5:05 PM, Eric Niebler <notifications@github.com>
wrote:

> We'll just pretend that never happened.
>
> LOL
>
> I think it will go away if we decide to go the range_traits route.
>
> I'm less enamored of this solution than I am with the separate
> is_sized_range trait. With the trait blob, it's possible for
> range_traits<R>::is_sized to be true even if size(r) doesn't compile.
> With the separate trait, that's not the case. And I'd like to investigate
> whether the hard error is actually a compiler bug. I can't make sense of it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ericniebler/stl2/issues/67#issuecomment-125766863>.
>



-- 
Casey Carter

**Comment by ericniebler** [comment]
> but it's clearly not an issue of great importance.

This. We can reevaluate if we find a more plausible scenario. The one you wrote just makes me think, "Don't do that."

**Comment by ericniebler** [comment]
> I think that hard errors are appropriate instead of SFINAE when the substitution depends on the completeness of a type.

In general yes because you don't want different compilation units to do different things if the type is complete in one place but not the other. But `void` is different. It's *never* complete. That said, I understand that it might be nigh impossible to make a compiler realize that.

**Comment by CaseyCarter** [comment]
Interestingly, `sizeof(incomplete type)` is not a hard error:
```
#include <type_traits>

template <class T, class E = void>
struct bar : std::false_type {};
template <class T>
struct bar<T, std::void_t<decltype(sizeof(T))>> : std::true_type {};

int main() {
    struct A;
    static_assert(!bar<A>());
}
```
[Compiles just fine.](http://melpon.org/wandbox/permlink/di0qJ1iDgTUxydAA) But now I'm rambling.

**Comment by ericniebler** [comment]
I'm merging this with the understanding that the resolution of #71 might change the design yet again.

**Comment by ericniebler** [comment]
Fixed in a8d8083
